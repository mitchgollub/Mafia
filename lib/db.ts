import mysql from 'serverless-mysql';
import { QueryError, RowDataPacket } from 'mysql';
import { SQLStatement } from 'sql-template-strings';

const db = mysql({
  config: {
    host: process.env.MYSQL_HOST,
    database: process.env.MYSQL_DATABASE,
    user: process.env.MYSQL_USER,
    password: process.env.MYSQL_PASSWORD,
  },
  onConnectError: (e: QueryError) => {
    console.error(`MySQL Connection Error: ${e.code}`);
  },
  onError: (e: QueryError) => {
    console.error(`MySQL Error: ${e.code}`);
  },
});

export async function query(query: SQLStatement): Promise<RowDataPacket[]> {
  const results: RowDataPacket[] | MySQLError = await db.query(query);
  await db.end();

  if ((results as MySQLError).error) {
    throw new Error(results.toString());
  }

  return results as RowDataPacket[];
}

interface MySQLError {
  error: string;
}
