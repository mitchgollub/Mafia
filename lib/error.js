const Error = {}

exports.InternalServerError = (res,error) => {
  console.error(`ERROR: ${error}`)
  res.status(500).json({
    error: error
  })
  return res;
}
