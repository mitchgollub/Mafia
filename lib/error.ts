import { NextApiResponse } from 'next';

function setStatus(
  res: NextApiResponse,
  code: number,
  message: string,
): NextApiResponse {
  console.error(message);
  res.status(code).json({
    error: code,
  });
  return res;
}

export function InternalServerError(
  res: NextApiResponse,
  error: Error,
): NextApiResponse {
  console.error(error.stack);
  return setStatus(res, 500, `Internal Server Error: ${error}`);
}
export function BadRequest(
  res: NextApiResponse,
  error: string,
): NextApiResponse {
  return setStatus(res, 400, `Bad Request: ${error}`);
}
