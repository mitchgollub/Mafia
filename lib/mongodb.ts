import MongoClient from 'mongodb';
import Game from '../models/game';
import sanitize from 'mongo-sanitize';

const mongoHost = process.env.MONGO_HOST as string;
const mongoDb = process.env.MONGO_DB as string;

export class MongoDb {
  async insertGameDocument(document: Game): Promise<Game> {
    const client = await this.connect();
    const collection = client.db(mongoDb).collection('games');

    const result = await collection.insertOne(sanitize(document));

    client.close();
    return document;
  }

  async findGameDocument(query: string): Promise<Game | null> {
    const client = await this.connect();
    const collection = client.db(mongoDb).collection('games');
    const result = await collection.findOne({
      code: sanitize(query.toUpperCase()),
    });
    client.close();

    if (result) {
      return new Game(result);
    }

    return null;
  }

  async updateGameDocument(game: Game): Promise<Game> {
    const client = await this.connect();
    const collection = client.db(mongoDb).collection('games');
    const result = await collection.updateOne(
      { code: sanitize(game?.code.toUpperCase()) },
      { $set: sanitize(game) },
    );

    client.close();

    return game;
  }

  private async connect(): Promise<MongoClient.MongoClient> {
    return await MongoClient.connect(mongoHost, {
      useUnifiedTopology: true,
    });
  }
}
