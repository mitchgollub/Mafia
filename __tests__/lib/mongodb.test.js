import MongoClient from 'mongodb';
import sanitize from 'mongo-sanitize';
import { MongoDb } from '../../lib/mongodb';
import Game from '../../models/game';

jest.mock('mongodb');
jest.mock('mongo-sanitize');

let mongoDb;

beforeAll(async () => {
  MongoClient.connect.mockResolvedValue({
    db: jest.fn().mockReturnValue({
      collection: jest.fn().mockReturnValue({
        insertOne: jest.fn().mockResolvedValue({}),
        findOne: jest.fn().mockResolvedValue(),
        updateOne: jest.fn().mockResolvedValue({}),
      }),
    }),
    close: jest.fn(),
  });

  sanitize.mockImplementation((request) => request);
  mongoDb = new MongoDb();
});

test('insertGameDocument returns query results', async () => {
  const actual = await mongoDb.insertGameDocument(new Game());

  expect(actual).toMatchSnapshot();
});

test('findGameDocument returns query results', async () => {
  const actual = await mongoDb.findGameDocument('123');

  expect(actual).toMatchSnapshot();
});

test('updateGameDocument returns query results', async () => {
  const actual = await mongoDb.updateGameDocument(new Game());

  expect(actual).toMatchSnapshot();
});
