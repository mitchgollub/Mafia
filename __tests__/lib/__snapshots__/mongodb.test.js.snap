// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findGameDocument returns query results 1`] = `null`;

exports[`insertGameDocument returns query results 1`] = `
Game {
  "code": "",
  "players": Object {
    "available": Array [],
    "current": Array [],
  },
}
`;

exports[`updateGameDocument returns query results 1`] = `
Game {
  "code": "",
  "players": Object {
    "available": Array [],
    "current": Array [],
  },
}
`;
